go-controller/pkg/kube/annotator.go

```diff
@@ -0,0 +1,102 @@
+package kube
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	kapi "k8s.io/api/core/v1"
+)
+
+// Annotator represents the exported methods for handling node annotations
+type Annotator interface {
+	Set(key string, value interface{}) error
+	SetWithFailureHandler(key string, value interface{}, failFn FailureHandlerFn) error
+	Del(key string)
+	Run() error
+}
+
+// FailureHandlerFn is a function called when adding an annotation fails
+type FailureHandlerFn func(node *kapi.Node, key string, val interface{})
+
+type action struct {
+	key    string
+	val    interface{}
+	failFn FailureHandlerFn
+}
+
+type nodeAnnotator struct {
+	kube Interface
+	node *kapi.Node
+
+	changes map[string]*action
+}
+
+// NewNodeAnnotator returns a new annotator for Node objects
+func NewNodeAnnotator(kube Interface, node *kapi.Node) Annotator {
+	return &nodeAnnotator{
+		kube:    kube,
+		node:    node,
+		changes: make(map[string]*action),
+	}
+}
+
+func (na *nodeAnnotator) Set(key string, val interface{}) error {
+	return na.SetWithFailureHandler(key, val, nil)
+}
+
+func (na *nodeAnnotator) SetWithFailureHandler(key string, val interface{}, failFn FailureHandlerFn) error {
+	if val != nil && reflect.TypeOf(val).Kind() != reflect.String {
+		// Marshal to a string value which is what annotations require
+		bytes, err := json.Marshal(val)
+		if err != nil {
+			return fmt.Errorf("failed to marshal %q value %v to string: %v", key, val, err)
+		}
+		val = string(bytes)
+	}
+
+	na.changes[key] = &action{
+		key:    key,
+		val:    val,
+		failFn: failFn,
+	}
+
+	return nil
+}
+
+func (na *nodeAnnotator) Del(key string) {
+	na.changes[key] = &action{key: key}
+}
+
+func (na *nodeAnnotator) Run() error {
+	changes := make(map[string]interface{})
+	for k, act := range na.changes {
+		// Ignore annotations that already exist with the same value
+		if existing, ok := na.node.Annotations[k]; existing != act.val || !ok {
+			changes[k] = act.val
+		}
+	}
+	if len(changes) == 0 {
+		return nil
+	}
+
+	if err := na.kube.SetAnnotationsOnNode(na.node, changes); err != nil {
+		// Let failure handlers clean up
+		for _, act := range na.changes {
+			if act.failFn != nil {
+				act.failFn(na.node, act.key, act.val)
```

Comment:
confusing that `val` is potentially not the same value the caller passed to `SetWithFailureHandler`.
also, that if you do

    na.SetWithFailureHandler("foo", "bar", ...)
    na.Del("foo")

the failure handler will be called even though we ignored that value. (Also, that there's no `DeleteWithFailureHandler`, really...)

Reply:
@danwinship Fixed the first part. But I don't think the second part is correct; the Delete() will replace the Set() in the changes map, so the failure handler will not be called.